### ISSUE TITLE
ts80007 is incorrect at times?

### ISSUE THREAD

**Issue by ORESoftware** [microsoft/TypeScript#58126]
### 🔍 Search Terms

async/await conditional is-promise promise

### ✅ Viability Checklist

- [X] This wouldn't be a breaking change in existing TypeScript/JavaScript code
- [X] This wouldn't change the runtime behavior of existing JavaScript code
- [X] This could be implemented without emitting different JS based on the types of the expressions
- [X] This isn't a runtime feature (e.g. library functionality, non-ECMAScript syntax with JavaScript output, new syntax sugar for JS, etc.)
- [X] This isn't a request to add a new utility type: https://github.com/microsoft/TypeScript/wiki/No-New-Utility-Types
- [X] This feature would agree with the rest of our Design Goals: https://github.com/Microsoft/TypeScript/wiki/TypeScript-Design-Goals

### ⭐ Suggestion

For reference, isPromise(x) looks like:

```js
(client as any).ensure = () => {
  if (client.isOpen) {
    return;
  }
  return client.connect();
}

const isPromise = (x: any) => {
  return x && typeof x.then === 'function' && typeof x.catch === 'function';
}
```

here is what I saw:
<kbd>
![image](https://github.com/microsoft/TypeScript/assets/11139560/dd4b6337-1150-4215-8073-c52696ace4c3)
</kbd>

this is pretty handy code, IMO, more performant etc, big win, but at the same time we all like to avoid if/else where possible, in this case, I think TS shouldn't complain?

### 📃 Motivating Example

see above

### 💻 Use Cases

1. What do you want to use this for?  async/await optimizations
2. What shortcomings exist with current approaches?
3. What workarounds are you using in the meantime?

**Comment by fatcerberus** [comment]
This one isn't a compile error though, is it?  I thought it was just a lightbulb in the IDE and you're free to ignore it.

**Comment by andrewbranch** [comment]
Just write the return type of `isPromise` as `x is Promise<any>` if you want to make the types accurate enough for this suggestion diagnostic to follow
